src/Coordination/Changelog.cpp

```diff
@@ -813,37 +823,42 @@ void Changelog::readChangelogAndInitWriter(uint64_t last_commited_log_index, uin
         assert(last_log_read_result != std::nullopt);
         assert(!existing_changelogs.empty());
 
-        /// Actually they shouldn't exist, but to be sure we remove them
-        removeAllLogsAfter(last_log_read_result->log_start_index);
-
-        /// This log, even if it finished with error shouldn't be removed
-        assert(existing_changelogs.find(last_log_read_result->log_start_index) != existing_changelogs.end());
-        assert(existing_changelogs.find(last_log_read_result->log_start_index)->first == existing_changelogs.rbegin()->first);
-
         /// Continue to write into incomplete existing log if it didn't finish with error
         const auto & description = existing_changelogs[last_log_read_result->log_start_index];
 
-        if (last_log_read_result->last_read_index == 0 || last_log_read_result->error) /// If it's broken log then remove it
+        const auto remove_invalid_logs = [&]
+        {
+            /// Actually they shouldn't exist, but to be sure we remove them
+            removeAllLogsAfter(last_log_read_result->log_start_index);
+
+            /// This log, even if it finished with error shouldn't be removed
+            chassert(existing_changelogs.find(last_log_read_result->log_start_index) != existing_changelogs.end());
+            chassert(existing_changelogs.find(last_log_read_result->log_start_index)->first == existing_changelogs.rbegin()->first);
+        };
+
+        if (last_log_read_result->last_read_index == 0) /// If it's broken or empty log then remove it
         {
-            LOG_INFO(log, "Removing chagelog {} because it's empty or read finished with error", description->path);
+            LOG_INFO(log, "Removing chagelog {} because it's empty", description->path);
+            remove_invalid_logs();
             description->disk->removeFile(description->path);
             existing_changelogs.erase(last_log_read_result->log_start_index);
             std::erase_if(logs, [last_log_read_result](const auto & item) { return item.first >= last_log_read_result->log_start_index; });
         }
+        else if (last_log_read_result->error)
```

Comment:
The main fix is here, right? Now if we have error we will not remove broken log.

Reply:
yes, and we continue reading if we detect error
https://github.com/ClickHouse/ClickHouse/pull/56670/files#diff-f1188c18e99d4fe945c6a979a345bb79c30a372c2c3b39524f312f70649a0648L766

gaps are still caught with the check that the differences between logs cannot be more than 1